# Week 8 Instructor lab

**Course Information:**  Graduate Introduction to Python, BGEN 632

**Notebook Author Name:** Leo Valladares

**Submission Date:**  2025/04/18

**Description:** This Jupyter notebook uses pandas' functions to do data manipulation

### Import Modules

In [3]:
import os
import pandas as pd
import numpy as np

### Set Working Directory

In [5]:
os.chdir("C:/Users/leova/OneDrive/BGEN 632/week8labs/data")

### Access Data

In [7]:
# Open the file within Python
gh_turnover = pd.read_csv("gh_turnover_four_projects.csv")
gh_turnover.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 473 entries, 0 to 472
Data columns (total 16 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   project_id     473 non-null    int64  
 1   domain         473 non-null    object 
 2   language       473 non-null    object 
 3   windows        473 non-null    int64  
 4   window_idx     473 non-null    int64  
 5   num_team       473 non-null    int64  
 6   num_commits    473 non-null    int64  
 7   blau_gender    473 non-null    float64
 8   Gini_gh_ten    473 non-null    float64
 9   user_id        473 non-null    int64  
 10  Core1          473 non-null    int64  
 11  commits        473 non-null    int64  
 12  propCommits    473 non-null    float64
 13  gender         449 non-null    object 
 14  github_tenure  473 non-null    int64  
 15  leavesNextQ    473 non-null    int64  
dtypes: float64(3), int64(10), object(3)
memory usage: 59.3+ KB


## Rename columns

In [9]:
(gh_turnover
 .rename({'domain':'project_domain',
         'language':'project_language',
         'windows':'project_age',
         'window_idx':'quarter',
         'num_team':'team_size',
         'num_commits':'project_commits',
         'blau_gender':'gender_blau',
         'Gini_gh_ten':'gh_tenure_gini',
         'Core1':'core_dev',
         'commits':'user_commits',
         'propCommits':'commit_proportion',
         'github_tenure':'gh_tenure',
         'leavesNextQ':'leaves_next_q'},  
        axis = 1,
        inplace = True) # Rename columns
)
gh_turnover.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 473 entries, 0 to 472
Data columns (total 16 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   project_id         473 non-null    int64  
 1   project_domain     473 non-null    object 
 2   project_language   473 non-null    object 
 3   project_age        473 non-null    int64  
 4   quarter            473 non-null    int64  
 5   team_size          473 non-null    int64  
 6   project_commits    473 non-null    int64  
 7   gender_blau        473 non-null    float64
 8   gh_tenure_gini     473 non-null    float64
 9   user_id            473 non-null    int64  
 10  core_dev           473 non-null    int64  
 11  user_commits       473 non-null    int64  
 12  commit_proportion  473 non-null    float64
 13  gender             449 non-null    object 
 14  gh_tenure          473 non-null    int64  
 15  leaves_next_q      473 non-null    int64  
dtypes: float64(3), int64(10), 

## Data Wrangling

In [11]:
# Calculate the mean of `user_commits`
(gh_turnover
 .filter(["user_commits"])
 .agg(["mean"])
)

user_commits
mean    339.365751

In [12]:
# Calculate the median of `user_commits`
(gh_turnover
 .filter(["user_commits"])
 .agg(["median"])
)

user_commits
median          18.0

In [13]:
# Select all columns that start with a *p* (i.e., `(^p)`) **or** contain an *g* (i.e., `(g)`). Save it as a new DataFrame named `turnover_new`. Output the columns in your notebook.
turnover_new = (gh_turnover
          .filter(regex = "(^p)|(g)")  # select desired columns
         )

turnover_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 473 entries, 0 to 472
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   project_id        473 non-null    int64  
 1   project_domain    473 non-null    object 
 2   project_language  473 non-null    object 
 3   project_age       473 non-null    int64  
 4   project_commits   473 non-null    int64  
 5   gender_blau       473 non-null    float64
 6   gh_tenure_gini    473 non-null    float64
 7   gender            449 non-null    object 
 8   gh_tenure         473 non-null    int64  
dtypes: float64(2), int64(4), object(3)
memory usage: 33.4+ KB


In [14]:
# Using your newly created DataFrame `p_or_g`, select rows in which `project_commits` is greater than 9,000.
(turnover_new
 .query("project_commits > 9000") # keep rows based on value the criteria 
 )

project_id project_domain project_language  project_age  project_commits  \
0          1101          OTHER           Python           23            22688   
1          1101          OTHER           Python           23            22688   
2          1101          OTHER           Python           23            22688   
3          1101          OTHER           Python           23            22688   
4          1101          OTHER           Python           23            22688   
..          ...            ...              ...          ...              ...   
116        1101          OTHER           Python           23            14210   
117        1101          OTHER           Python           23            14210   
118        1101          OTHER           Python           23            14210   
119        1101          OTHER           Python           23            14210   
120        1101          OTHER           Python           23            14210   

     gender_blau  gh_tenure_gini gender  gh_tenure  
0            0.0        0.422259   male        426  
1            0.0        0.422259   male         46  
2            0.0        0.422259   male        941  
3            0.0        0.422259   male       1870  
4            0.0        0.422259   male        781  
..           ...             ...    ...        ...  
116          0.0        0.411372   male        495  
117          0.0        0.411372   male        487  
118          0.0        0.411372   male        317  
119          0.0        0.411372   male        418  
120          0.0        0.411372   male        179  

[114 rows x 9 columns]

In [15]:
# Using pandas piping notation, perform the previous two operations together and save it as a new DataFrame `turnover_newer`. 
# This means you should select columns that start with a *p* or contain an *g* and select rows in which `project_commits` is greater than 9,000
turnover_newer = (gh_turnover
                  .filter(regex = "(^p)|(g)")  # select desired columns
                  .query("project_commits > 9000") # keep rows based on value the criteria 
                 )
turnover_newer

project_id project_domain project_language  project_age  project_commits  \
0          1101          OTHER           Python           23            22688   
1          1101          OTHER           Python           23            22688   
2          1101          OTHER           Python           23            22688   
3          1101          OTHER           Python           23            22688   
4          1101          OTHER           Python           23            22688   
..          ...            ...              ...          ...              ...   
116        1101          OTHER           Python           23            14210   
117        1101          OTHER           Python           23            14210   
118        1101          OTHER           Python           23            14210   
119        1101          OTHER           Python           23            14210   
120        1101          OTHER           Python           23            14210   

     gender_blau  gh_tenure_gini gender  gh_tenure  
0            0.0        0.422259   male        426  
1            0.0        0.422259   male         46  
2            0.0        0.422259   male        941  
3            0.0        0.422259   male       1870  
4            0.0        0.422259   male        781  
..           ...             ...    ...        ...  
116          0.0        0.411372   male        495  
117          0.0        0.411372   male        487  
118          0.0        0.411372   male        317  
119          0.0        0.411372   male        418  
120          0.0        0.411372   male        179  

[114 rows x 9 columns]

In [16]:
# Using pandas piping notation, select all columns that end with the letter *e* (i.e., `e$`) or contain the letter *q* (i.e., `s`). 
# Additionally, select rows in which `gh_tenure` is greater than or equal to 1,991. 
(gh_turnover
 .filter(regex = "e$|s")  # select desired columns
 .query("gh_tenure >= 1991") # keep rows based on value the criteria     
)


project_language  project_age  team_size  project_commits  user_id  \
339             Java           16          2                7    33272   
457           Python           23         20             3070    88459   

     user_commits  gh_tenure  leaves_next_q  
339             6       2006              0  
457             1       2054              0

In [17]:
# Use the query you just performed. You will calculate the mean and median of `user_commits`. 
(gh_turnover
 .filter(regex = "e$|s")  # select desired columns
 .query("gh_tenure >= 1991") # keep rows based on value the criteria    
 .filter(["user_commits"])
 .agg(["mean","median"])
)

user_commits
mean             3.5
median           3.5

#### How does it compare to the answers above in which you calculated the mean and median?

In the previous calculation with 473 observations, with a mean of 339.4 and a median of 18, the data showed strong right skew. This could be due to the presence of large outliers. In contrast, the current calculation yields a mean and median of 3.5, based on only two observations with user_commit values of 1 and 6. This smaller dataset does not have the extreme values that previously increased the mean, resulting in a more balanced summary, but with a small sample that could not represent the overall population. 
